proposals/function-pointers.md

```diff
@@ -0,0 +1,245 @@
+exactly what signature they are emitted with.
+
+### Better function member
+The better function member specification will be changed to include the following line:
+
+> A `funcptr` is more specific than `void*`
+
+This means that it is possible to overload on `void*` and a `funcptr` and still sensibly use the address-of operator.
+
+## Open Issuess
+
+- Round tripping function pointer names, as well as parameter names, through metadata will require additional work. The
+function pointer type itself is natively supported by CLI but that does not include any names. This is not anticipated
+to be a big issue, just needs design work.
+- The address-of operator is limited to `static` methods in this proposal. It can be made to work with instance methods 
+but the behavior can be confusing to developers. The `this` type becomes an explicit first parameter on the `funcptr` 
+type. This means the behavior and usage would differ significantly from `delegate`. This extra confusion was the main 
+reason it was not included in the design.
+
+## Considerations
+
+### Don't require unsafe at declaration
+Instead of requiring `unsafe` at every use of a `funcptr`, only require it at the point where a method group is
+converted to a `funcptr`. This is where the core safety issues come into play (knowing that the containing assembly
+cannot be unloaded while the value is alive). Requiring `unsafe` on the other locations can be seen as excessive.
+
+This is how the design was originally intended. But the resulting language rules felt very awkward. It's impossible to
+hide the fact that this is a pointer value and it kept peeking through even without the `unsafe` keyword. For example 
+the conversion to `object` can't be allowed, it can't be a member of a `class`, etc ... The C# design is to require
+`unsafe` for all pointer uses and hence this design follows that.
+
+Developers will still be capable of preventing a _safe_ wrapper on top of `funcptr` values the same way that they do 
+for normal pointer types today. Consider:
+
+``` csharp
+unsafe struct Action {
+    funcptr void() _ptr;
+
+    Action(funcptr void() ptr) => _ptr = ptr;
+    public void Invoke() => _ptr();
+}
+```
+
+### Using delegates
+Instead of using a new syntax element, `funcptr`, simply use exisiting `delegate` types with a `*` following the type:
+
+``` csharp
+Func<object, object, bool>* ptr = &object.ReferenceEquals;
+```
+
+Handling calling convention can be done by annotating the `delegate` types with an attribute that specifies 
+a `CallingConvention` value. The lack of an attribute would signify the managed calling convention.
+
+Encoding this in IL is problematic. The underlying value needs to be represented as a pointer yet it also must:
+
+1. Have a unique type to allow for overloads with different function pointer types. 
+1. Be equivalent for OHI purposes across assembly boundaries.
```

Comment:
What's OHI? I found it in https://github.com/dotnet/roslyn/issues/20528, where it means "override, hiding, implementing", but I think it's not a familiar acronym and so it should be expanded here.

Reply:
OHI is the term to describe how method signature are compared when we are doing interface implementation, override and member hiding hookup. More specifically:

``` csharp
class Base {
  protected virtual void M(int i) { ... }
}
class Derived : Base { 
  protected override void M(int j) { ... }
}
```

The rules for determining if `Deride.M` is a proper override for `Base.M` is generally referred to as OHI.